### PR TITLE
[LMS-116] Load section association files

### PR DESCRIPTION
**Pull request by bradbanister**

### CODE REVIEW

File: docs/metaed-project/Association/UserLMSSectionAssociation.metaed [hunk 1]
```diff
@@ -25,10 +25,10 @@ Association LMSUserLMSSectionAssociation
         role name Enrollment
     date StartDate
         documentation "Month, day, and year of the user's entry or assignment to the section."
-        is required
+        is optional
```
Comment (bradbanister) [inline docs/metaed-project/Association/UserLMSSectionAssociation.metaed:28]: Made this change because all three extractors have put nulls for StartDate/EndDate.  Will note in ticket for discussion.

File: src/file-utils/edfi_lms_file_utils/file_reader.py [hunk 2]
```diff
@@ -268,7 +268,21 @@ def read_section_associations_file(
     -------
     Pandas DataFrame with columns matching the model definition / CSV file.
     """
-    return _read_csv(full_path, nrows, log_level=logging.DEBUG)
+    data_types = {
```
Comment (bradbanister) [inline src/file-utils/edfi_lms_file_utils/file_reader.py:271]: Updating data types as suggested.

File: src/lms-ds-loader/edfi_lms_ds_loader/loader_facade.py [hunk 3]
```diff
@@ -105,9 +105,11 @@ def __get_all_file_paths_callback():
 
 
 def _load_assignments(csv_path: str, db_adapter: MssqlLmsOperations) -> None:
-    sections = set(
-        file_reader.get_all_sections(csv_path)["SourceSystemIdentifier"]
-    )  # we only need to access the last file
+    sections_df = file_reader.get_all_sections(csv_path)
```
Comment (bradbanister) [inline src/lms-ds-loader/edfi_lms_ds_loader/loader_facade.py:108]: Be polite instead of crashing if section df is empty, probably because of wrong csv path.  Not that I learned this the hard way...

File: src/lms-ds-loader/tests/test_df_to_db.py [hunk 4]
```diff
@@ -15,9 +15,9 @@
 SOURCE_SYSTEM = "google"
 
 
-def describe_given_a_resource_that_is_not_a_child_of_section() -> None:
+def describe_given_a_resource_that_is_not_a_child_of_section_or_user() -> None:
     @pytest.fixture
-    def when_uploading_assignments() -> Tuple[MagicMock, pd.DataFrame]:
+    def when_uploading_users() -> Tuple[MagicMock, pd.DataFrame]:
```
Comment (bradbanister) [inline src/lms-ds-loader/tests/test_df_to_db.py:20]: Actually a test of user, probably copy-paste thing.

File: src/lms-ds-loader/tests/test_loader_facade.py [hunk 5]
```diff
@@ -58,23 +58,34 @@ def fixture(mocker) -> Tuple[Dict[str, MagicMock], Dict[str, pd.DataFrame]]:
...
 
+            fake_df_section_associations = pd.DataFrame([{"assignments": "b"}])
+            mocker.patch(
+                "edfi_lms_file_utils.file_reader.read_section_associations_file",
+                return_value=fake_df_section_associations,
+            )
+
             mock_upload_file = mocker.patch("edfi_lms_ds_loader.df_to_db.upload_file")
-            mock_upload_submissions_file = mocker.patch(
+            mock_upload_assignments_file = mocker.patch(
```
Comment (bradbanister) [inline src/lms-ds-loader/tests/test_loader_facade.py:68]: Fixed typo here.

File: docs/metaed-project/Association/UserLMSSectionAssociation.metaed (hunk 1 above)
Comment (stephenfuqua) [reply docs/metaed-project/Association/UserLMSSectionAssociation.metaed:28]: Interesting, if none of them support a StartDate or EndDate for a section association, why do we have these columns? Or were we going to infer this? Definitely worth an extra discussion about whether they should be kept.

File: src/file-utils/edfi_lms_file_utils/file_reader.py [hunk 6]
```diff
@@ -268,7 +268,21 @@ def read_section_associations_file(
...
+        "LMSSectionSourceSystemIdentifier": "string",
+        "LMSUserSourceSystemIdentifier": "string",
+    }
+
+    extra_date_columns = ["StartDate", "EndDate"]
+
+    return _read_csv(
+        full_path,
+        nrows,
+        log_level=logging.DEBUG,
```
Comment (stephenfuqua) [inline src/file-utils/edfi_lms_file_utils/file_reader.py:282]: Why hard-coding the log level to DEBUG?

File: src/lms-ds-loader/edfi_lms_ds_loader/df_to_db.py [hunk 7]
```diff
@@ -115,3 +115,21 @@ def upload_assignments(
...
+    TABLE = Table.SECTION_ASSOCIATION
+    columns = list(section_associations_df.columns)
+
+    _prepare_staging_table(db_adapter, section_associations_df, TABLE)
+    db_adapter.insert_new_records_to_production_for_section_and_user(
+        TABLE,
+        columns
+    )
+    db_adapter.copy_updates_to_production(TABLE, columns)
+    db_adapter.soft_delete_from_production(TABLE, _get_source_system(section_associations_df))
```
Comment (stephenfuqua) [inline src/lms-ds-loader/edfi_lms_ds_loader/df_to_db.py:133]: Is there an opportunity to refactor this and `_upload_assignments` to reduce duplicated code?

File: src/lms-ds-loader/edfi_lms_ds_loader/loader_facade.py (hunk 3 above)
Comment (stephenfuqua) [reply src/lms-ds-loader/edfi_lms_ds_loader/loader_facade.py:108]: Is there a way to refactor and pass the `sections_df` into these functions, so that we only read once?

File: src/lms-ds-loader/edfi_lms_ds_loader/mssql_lms_operations.py [hunk 8]
```diff
@@ -218,6 +218,74 @@ def insert_new_records_to_production_for_section(
...
+        assert len(columns) > 0, "Argument `columns` cannot be empty"
+
+        insert_columns = ",".join(
+            [f"\n    {c}" for c in columns if not (c == "LMSSectionSourceSystemIdentifier" or c == "LMSUserSourceSystemIdentifier")]
+        )
+        select_columns = ",".join(
+            [
+                f"\n    stg.{c}"
+                for c in columns
+                if not (c == "LMSSectionSourceSystemIdentifier" or c == "LMSUserSourceSystemIdentifier")
```
Comment (stephenfuqua) [inline src/lms-ds-loader/edfi_lms_ds_loader/mssql_lms_operations.py:251]: Only while doing the assignments upload did I realize how closely list comprehensions mirror SQL statements - when I had this kind of formatting.

File: src/lms-ds-loader/tests/test_mssql_lms_operations.py [hunk 9]
```diff
@@ -185,6 +185,37 @@ def it_should_build_a_valid_insert_statement(mocker) -> None:
             # Assert
             exec_mock.assert_called_with(expected)
 
+    def describe_given_resource_is_child_of_section_and_user() -> None:
+        def it_should_build_a_valid_insert_statement(mocker) -> None:
```
Comment (stephenfuqua) [inline src/lms-ds-loader/tests/test_mssql_lms_operations.py:189]:
```suggestion
        def it_should_build_a_valid_update_statement(mocker) -> None:
```

**Review by stephenfuqua — CHANGES_REQUESTED** [review]

File: src/lms-ds-loader/edfi_lms_ds_loader/loader_facade.py (hunk 3 above)
Comment (bradbanister) [reply src/lms-ds-loader/edfi_lms_ds_loader/loader_facade.py:108]: I'll memoize it. 😎

File: src/file-utils/edfi_lms_file_utils/file_reader.py (hunk 6 above)
Comment (bradbanister) [reply src/file-utils/edfi_lms_file_utils/file_reader.py:282]: Good question, that's not a change of mine though. Just an auto reformat.

Comment (lalrg) [reply src/file-utils/edfi_lms_file_utils/file_reader.py:282]: I did it, the _read_csv has a default info level... Setting it to debug level was part of the acceptance criteria for the" do not reprocess files" ticket. The reason was to don't get a lot of messages in the log.

File: src/lms-ds-loader/edfi_lms_ds_loader/loader_facade.py (hunk 3 above)
Comment (lalrg) [reply src/lms-ds-loader/edfi_lms_ds_loader/loader_facade.py:108]: I was creating a shared datastore because I am having the same need. 😆

File: src/lms-ds-loader/edfi_lms_ds_loader/migrator.py [hunk 10]
```diff
@@ -23,6 +23,7 @@
     "create_user_tables",
     "create_section_tables",
     "create_assignment_tables",
+    "create_section_association_tables",
```
Comment (lalrg) [inline src/lms-ds-loader/edfi_lms_ds_loader/migrator.py:26]: You can put it after create_processed_files_table 🤔

File: src/lms-ds-loader/edfi_lms_ds_loader/mssql_lms_operations.py (hunk 8 above)
Comment (lalrg) [reply src/lms-ds-loader/edfi_lms_ds_loader/mssql_lms_operations.py:251]: 🤯

File: src/lms-ds-loader/edfi_lms_ds_loader/df_to_db.py (hunk 7 above)
Comment (bradbanister) [reply src/lms-ds-loader/edfi_lms_ds_loader/df_to_db.py:133]: I've got a nice refactor of this, through it's made testing with mocks a little more interesting...

File: src/file-utils/edfi_lms_file_utils/file_reader.py (hunk 6 above)
Comment (stephenfuqua) [reply src/file-utils/edfi_lms_file_utils/file_reader.py:282]: I overlooked this in the prior review. The code should have used `logger.Debug(message)` instead of `logger.log(log_level, message)`. Its not a big deal, but perhaps we can opportunistically refactor that in the future.

File: src/lms-ds-loader/edfi_lms_ds_loader/migrator.py (hunk 10 above)
Comment (stephenfuqua) [reply src/lms-ds-loader/edfi_lms_ds_loader/migrator.py:26]: Yes, it is best to add the migration at the end of the list. But probably doesn't really matter in this case.

File: src/lms-ds-loader/edfi_lms_ds_loader/loader_facade.py (hunk 3 above)
Comment (bradbanister) [reply src/lms-ds-loader/edfi_lms_ds_loader/loader_facade.py:108]: The Python way of memoization: https://dbader.org/blog/python-memoization

TLDR:
`@lru_cache()
def _get_sections_df(csv_path: str) -> pd.DataFrame:
    return file_reader.get_all_sections(csv_path)`

Comment (bradbanister) [reply src/lms-ds-loader/edfi_lms_ds_loader/loader_facade.py:108]: Very common in functional programming.

Comment (stephenfuqua) [reply src/lms-ds-loader/edfi_lms_ds_loader/loader_facade.py:108]: Looks very interesting

File: src/lms-ds-loader/edfi_lms_ds_loader/migrator.py (hunk 10 above)
Comment (bradbanister) [reply src/lms-ds-loader/edfi_lms_ds_loader/migrator.py:26]: Given the stage of development we're at, I wasn't concerned beyond making sure relational dependencies are in order.  I can't imagine not dropping/recreating dbs from scratch at this point during testing.

Comment (stephenfuqua) [reply src/lms-ds-loader/edfi_lms_ds_loader/migrator.py:26]: Fair enough

File: src/lms-ds-loader/edfi_lms_ds_loader/df_to_db.py (hunk 7 above)
Comment (bradbanister) [reply src/lms-ds-loader/edfi_lms_ds_loader/df_to_db.py]: Was able to simplify and share with assignments, along with test simplification.

**Review by stephenfuqua — APPROVED** [review]
